metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/MetricQueryBuilder.java

```diff
@@ -0,0 +1,267 @@
+      String columnName, Operator op, long value) {
+    return org.hypertrace.core.query.service.api.Filter.newBuilder()
+        .setLhs(createColumnExpression(columnName))
+        .setOperator(op)
+        .setRhs(createLongLiteralExpression(value))
+        .build();
+  }
+
+  public static Expression createLongLiteralExpression(long value) {
+    return Expression.newBuilder()
+        .setLiteral(
+            LiteralConstant.newBuilder()
+                .setValue(Value.newBuilder().setValueType(ValueType.LONG).setLong(value)))
+        .build();
+  }
+
+  static void setSelection(
+      QueryRequest.Builder queryBuilder,
+      Attribute metricAttribute,
+      MetricAggregationFunction metricAggregationFunction) {
+    Function function =
+        Function.newBuilder()
+            .setFunctionName(convertToQueryRequestFunctionName(metricAggregationFunction))
+            .addArguments(
+                Expression.newBuilder()
+                    .setColumnIdentifier(
+                        ColumnIdentifier.newBuilder()
+                            .setColumnName(
+                                metricAttribute.getScope() + "." + metricAttribute.getKey())
+                            .build()))
+            .build();
+    queryBuilder.addSelection(Expression.newBuilder().setFunction(function));
+  }
+
+  static String convertToQueryRequestFunctionName(
+      MetricAggregationFunction metricAggregationFunction) {
+    switch (metricAggregationFunction) {
+      case METRIC_AGGREGATION_FUNCTION_TYPE_SUM:
+        return FunctionType.SUM.name();
+      case METRIC_AGGREGATION_FUNCTION_TYPE_AVG:
+        return FunctionType.AVG.name();
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  static Expression createTimeColumnGroupByExpression(String timeColumn, long periodSecs) {
+    return Expression.newBuilder()
+        .setFunction(
+            Function.newBuilder()
+                .setFunctionName(DATE_TIME_CONVERTER)
+                .addArguments(createColumnExpression(timeColumn))
+                .addArguments(createStringLiteralExpression("1:MILLISECONDS:EPOCH"))
+                .addArguments(createStringLiteralExpression("1:MILLISECONDS:EPOCH"))
+                .addArguments(createStringLiteralExpression(periodSecs + ":SECONDS")))
+        .build();
+  }
+
+  static Expression createColumnExpression(String columnName) {
+    return Expression.newBuilder()
+        .setColumnIdentifier(ColumnIdentifier.newBuilder().setColumnName(columnName))
+        .build();
+  }
+
+  static Expression createStringLiteralExpression(String value) {
+    return Expression.newBuilder()
+        .setLiteral(
+            LiteralConstant.newBuilder()
+                .setValue(Value.newBuilder().setValueType(ValueType.STRING).setString(value)))
+        .build();
+  }
+
+  String getTimestampAttributeId(String tenantId, String attributeScope) {
```

Comment:
Do you think, we should cache this - <scope, timeColumn>? As this will iterator for each rule execution?

Reply:
Yes, was to add this as part of independent pr